File: generic_programming/PetersNumberType.java
/**
* This is a short example of how a class template works in Java
**/


public class PetersNumberType<T> {

    private T n;

    public void setNum(T num) {
        this.n = num;
    }

    public T getNum() {
        return this.n;
    }

    public static void main(String[] args) {
        PetersNumberType<Integer> intNumber = new PetersNumberType<Integer>();
        intNumber.setNum(new Integer(42));

        PetersNumberType<Double> doubleNumber = new PetersNumberType<Double>();
        doubleNumber.setNum(new Double(3.1415));

        System.out.println(intNumber.getNum() + "is an int.");
        System.out.println("     ...but also a PetersNumberType.");

        System.out.println(doubleNumber.getNum() + "is a double.");
        System.out.println("     ...but also a PetersNumberType.");
    }
}

File: generic_programming/PetersNumberType.cpp
/**
* This is a short example of how a class template works in c+
**/

#include <iostream>
using namespace std;

template <class T>
class PetersNumberType {
    T n;
  public:
    PetersNumberType (T num)
        {n = num;}
    T getNum ();
};

template <class T>
T PetersNumberType<T>::getNum ()
{
    T returnMe;
    returnMe = n;
    return returnMe;
}

int main () {
    PetersNumberType <int> intNumber (42);

    PetersNumberType <double> doubleNumber (3.1415);

    cout << intNumber.getNum() << " is an int." << endl;
    cout << "     ...but also a PetersNumberType." << endl;

    cout << doubleNumber.getNum() << " is a double." << endl;
    cout << "     ...but it is also a PetersNumberType." << endl;

    return 0;
}